jax/core.py

```diff
@@ -1623,3 +1630,39 @@ def initial_style_staging():
       yield
     finally:
       trace_state.initial_style = prev
+
+class Float0Array:
+  """Array used to represent the tangent values of int/bool primals.
+
+  No operations are defined for this array-type by design: these values can't be
+  used in combination with non-trivial vector spaces. No implicit dtype
+  conversion is done.  To use these values, the array can be cast to float32
+  zeros using the `to_zeros` method.
+  """
```

Comment:
I have mixed feelings about the `Float0Array` class. I understand the motivation to explicitly not support any operations, and the desire to raise errors as close as possible to the source of the issue as possible, but I am concerned that this special case may cause more trouble than it's worth.

One concern is type stability/predictable. Currently, the return value from functions transformed with `grad` and `jit` is always a pytree of `DeviceArray` objects (but see https://github.com/google/jax/issues/4386). Now it's `DeviceArray` and/or `Float0Array` objects.

In contrast, the standard way to add new dtypes in NumPy/JAX is to just add a new dtype and use the standard array object, relying on the lack of operation implementations (e.g., for arithmetic) to raise errors, but still supporting methods like `transpose`/`reshape`, which are otherwise valid on arrays with any dtype.

A good example of this is the `jax.jacrev`, which despite being well-defined currently results in an error from inside `_unravel_array_into_pytree`:
```
In [29]: jax.jacrev(lambda x: 1.0)(1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-29-d57c9bab827f> in <module>
----> 1 jax.jacrev(lambda x: 1.0)(1)

~/dev/jax/jax/api.py in jacfun(*args, **kwargs)
    975     jac = jac[0] if isinstance(argnums, int) else jac
    976     example_args = dyn_args[0] if isinstance(argnums, int) else dyn_args
--> 977     jac = tree_map(partial(_unravel_array_into_pytree, y, 0), jac)
    978     return tree_transpose(tree_structure(example_args), tree_structure(y), jac)
    979

~/dev/jax/jax/tree_util.py in tree_map(f, tree)
    162   """
    163   leaves, treedef = pytree.flatten(tree)
--> 164   return treedef.unflatten(map(f, leaves))
    165
    166 def tree_multimap(f, tree, *rest):

~/dev/jax/jax/api.py in _unravel_array_into_pytree(pytree, axis, arr)
   1060   axis = axis % arr.ndim
   1061   shapes = [arr.shape[:axis] + np.shape(l) + arr.shape[axis+1:] for l in leaves]
-> 1062   parts = _split(arr, np.cumsum(map(np.size, leaves[:-1])), axis)
   1063   reshaped_parts = [np.reshape(x, shape) for x, shape in zip(parts, shapes)]
   1064   return tree_unflatten(treedef, reshaped_parts)

~/dev/jax/jax/api.py in _split(x, indices, axis)
   1068     return np.split(x, indices, axis)
   1069   else:
-> 1070     return x.split(indices, axis)
   1071
   1072 def _dtype(x):

~/dev/jax/jax/numpy/lax_numpy.py in float0_error(self, *args, **kwargs)
   4657 def float0_error(self: core.Float0Array, *args, **kwargs):
   4658   raise TypeError(
-> 4659       "Float0Arrays do not support any operations by design, because float0s "
   4660       "are not compatible with non-trivial vector spaces. No implicit dtype "
   4661       "conversion is done. You can use to_zeros() to cast this array to "

TypeError: Float0Arrays do not support any operations by design, because float0s are not compatible with non-trivial vector spaces. No implicit dtype conversion is done. You can use to_zeros() to cast this array to float32 zeros.
If you didn't expect to get a Float0Array you might have accidentally taken a gradient with respect to an integer or boolean argument.
```

You can try to fix all these cases inside JAX, but it's a bit of an uphill battle. Sticking to operations that are unsupported by a dtype would be a bit simpler.

Reply:
Thanks for the review Stephan, I agree this is not very neat currently. The only reason for having a separate array type was to provide more context on an error, but I think it's a good idea to do this at the dtype promotion level. It makes more sense in a way: what we're trying to prevent is implicit dtype casting because this could lead to silently doing the wrong thing. 

> the standard way to add new dtypes in NumPy/JAX is to just add a new dtype and use the standard array object

Which array object is this? DeviceArray? a numpy array? 

> As long as it's excluded from all dtype promotion, I imagine it could be perfectly fine (and maybe even useful) to support arbitrary operation like lax.add.

Would this mean only supporting this for float0+float0 cases because there is no promotion implemented?

Reply:
I guess one additional difference is a Float0Array does not actually store any values, it only defines a shape. Is there a way to express this using the existing arrays? Maybe a zeros array with float0 values which are defined as taking up 0 bytes is actually the same as not storing values?

In this way it's like a ShapedArray, but @mattjj explained to me we need an actual array next to our abstract arrays, so I don't think we can reuse ShapedArray?

Reply:
After some discussion we decided to abandon the Float0Array and opt for ndarrays with float0 dtype. This allows for dtype-agnostic operations (like reshape or split). The float0error is now thrown when promoting the dtype of the arguments in lax and lax.numpy.